File: core/src/main/java/com/mcreater/amclcore/command/script/ScriptManager.java
package com.mcreater.amclcore.command.script;

import com.mcreater.amclcore.command.CommandArg;

import java.util.*;

import static com.mcreater.amclcore.i18n.I18NManager.translatable;

public class ScriptManager {
    private static final Map<String, ScriptProvider> scriptProviders = new HashMap<>();

    static {
        register("powershell", PowerShellScriptProvider.Companion.getINSTANCE());
        register("cmdlet", CmdletScriptProvider.Companion.getINSTANCE());
        register("sh", ShellScriptProvider.Companion.getINSTANCE());
    }

    public static List<String> getProviders() {
        return new Vector<>(scriptProviders.keySet());
    }

    public static String getScriptDesc(String id) {
        return translatable(String.format("core.script.%s.desc", id)).getText();
    }

    public static String toScript(String id, List<CommandArg> args) {
        return Optional.ofNullable(scriptProviders.get(id)).map(sp -> sp.toScript(args)).orElse("");
    }

    public static void register(String id, ScriptProvider provider) {
        scriptProviders.put(id, provider);
    }
}

File: core/src/main/java/com/mcreater/amclcore/command/script/ScriptManager.kt
package com.mcreater.amclcore.command.script

import com.mcreater.amclcore.command.CommandArg
import com.mcreater.amclcore.i18n.I18NManager
import com.mcreater.amclcore.i18n.Text
import java.util.*


class ScriptManager {
    companion object {
        @JvmStatic
        private val scriptProviders: MutableMap<String, ScriptProvider> = HashMap()

        init {
            register("powershell", PowerShellScriptProvider.INSTANCE)
            register("cmdlet", CmdletScriptProvider.INSTANCE)
            register("sh", ShellScriptProvider.INSTANCE)
        }

        @JvmStatic
        val providers: List<String>
            get() = Vector(scriptProviders.keys)

        @JvmStatic
        fun getScriptDesc(id: String?): Text {
            return I18NManager.translatable(String.format("core.script.%s.desc", id))
        }

        @JvmStatic
        fun toScript(id: String, args: List<CommandArg?>?): String? {
            return Optional.ofNullable(scriptProviders[id]).map { sp: ScriptProvider ->
                sp.toScript(
                    args
                )
            }.orElse("")
        }

        @JvmStatic
        fun register(id: String, provider: ScriptProvider) {
            scriptProviders[id] = provider
        }
    }
}
